### PR TITLE
fix(AOBS-491): Use native sfType instead of originalType, and fix schema-less tables

### LINKED ISSUE #19: KeyError when Snowflake table has no schema
If a Snowflake table has no Schema (e.g., because it's in `PUBLIC`), the plugin fails with:

```
[2020/07/29-16:57:23.386] [null-err-102] [INFO] [dku.utils]  - *************** Recipe code failed **************
[2020/07/29-16:57:23.387] [null-err-102] [INFO] [dku.utils]  - Begin Python stack
[2020/07/29-16:57:23.387] [null-err-102] [INFO] [dku.utils]  - Traceback (most recent call last):
[2020/07/29-16:57:23.387] [null-err-102] [INFO] [dku.utils]  -   File "/mnt/secondary/dataiku_data/jobs/KLABERSB2/Build_dss_sf_date_test_hdfs_2020-07-29T16-57-12.834/compute_dss_sf_date_test_hdfs_NP/custom-python-recipe/pyouthmC6vTuQNYAr/python-exec-wrapper.py", line 194, in <module>
[2020/07/29-16:57:23.388] [null-err-102] [INFO] [dku.utils]  -     exec(f.read())
[2020/07/29-16:57:23.388] [null-err-102] [INFO] [dku.utils]  -   File "<string>", line 16, in <module>
[2020/07/29-16:57:23.388] [null-err-102] [INFO] [dku.utils]  -   File "/mnt/secondary/dataiku_data/plugins/dev/snowflake-hdfs/python-lib/config.py", line 54, in get_table_name
[2020/07/29-16:57:23.389] [null-err-102] [INFO] [dku.utils]  -     return f'"{params["schema"]}"."{params["table"]}"'.replace('${projectKey}', project_key)
[2020/07/29-16:57:23.389] [null-err-102] [INFO] [dku.utils]  - KeyError: 'schema'
```

This is because `get_config()["params"]` does not return a `schema` value:

```
{'connection': 'outra-sf-dataiku-db',
 'notReadyIfEmpty': False,
 'mode': 'table',
 'partitioningType': 'custom',
 'normalizeDoubles': True,
 'table': 'DSS_SF_DATE_TEST',
 'tableCreationMode': 'auto',
 'writeInsertBatchSize': 10000,
 'writeJDBCBadDataBehavior': 'DISCARD_ROW',
 'readColsWithUnknownTzAsDates': False,
 'readSQLDateColsAsDSSDates': True}
```

image

### PR DESCRIPTION
**Pull request by mklaber**
This PR fixes the issue with unscores in the `originalType` parameter while keeping true to the intent of `originalType`.

It also applies flake8 formatting rules.

Fixes #19 

Closes #18

### CODE REVIEW

File: python-lib/config.py [hunk 1]
```diff
@@ -84,19 +84,21 @@ def get_hdfs_location(dataset_config: Mapping[AnyStr, Any], sf_stage_name: AnySt
...
     """
     tab = sf_table_name.replace('"', '')
     schema = tab.split('.')[0]
     table = tab.split('.')[1]
 
-    sql = f'''
-SELECT column_name AS "name", data_type AS "originalType"
+    # Dataiku's `originalType` parameter is Snowflake's data_type but without `_`
+    sql = f"""
+SELECT column_name AS "name", REPLACE(data_type, '_', '') AS "originalType"
```
Comment (czen88) [inline python-lib/config.py:97]: What is the purpose for this change? This is going to break my PR, so not sure if this is correct.

File: python-lib/config.py [hunk 2]
```diff
@@ -121,8 +123,7 @@ def get_snowflake_to_hdfs_query(sf_location: AnyStr, sf_table_name: AnyStr,
     # This is required as the COPY command does not support TZ and LTZ
     timezoned_types = ['TIMESTAMPLTZ', 'TIMESTAMPTZ', 'TIMESTAMP']
     columns = [
-        f'"{c["name"]}"'
-        + (f'::TIMESTAMP_NTZ AS "{c["name"]}"' if c['originalType'] in timezoned_types else '')
+        f'"{c["name"]}"' + (f'::TIMESTAMP_NTZ AS "{c["name"]}"' if c['originalType'] in timezoned_types else '')
```
Comment (czen88) [inline python-lib/config.py:126]: This change is going to be in conflict with my PR. No need to handle TIMESTAMP type as internally it is replaced by actual type. As per [Snowflake documentation] (https://docs.snowflake.com/en/sql-reference/data-types-datetime.html) 'The TIMESTAMP data type is never stored in tables.'

Comment (mklaber) [reply python-lib/config.py:126]: IIRC, `['TIMESTAMPLTZ', 'TIMESTAMPTZ', 'TIMESTAMP']` were the observed data types from `read_schema()` in the instances where `read_schema` was actually populating `originalType`. Because this method is still acting upon what it'd get from `read_schema`, it should continue to cover the cases where it could theoretically end up getting `TIMESTAMP`. In other words: this method doesn't know that we've manually retrieved an `originalType` value directly from Snowflake.

File: python-lib/config.py (hunk 1 above)
Comment (mklaber) [reply python-lib/config.py:97]: This is instead of your PR, not in addition to.

File: python-lib/config.py (hunk 2 above)
Comment (czen88) [reply python-lib/config.py:126]: I do not know where the requirement to support Dataiku schema in this method is coming from. I understand that originally it was using one, but AFAIK this method is only used in one place so not sure what we are trying to achieve here.

File: python-lib/config.py (hunk 1 above)
Comment (czen88) [reply python-lib/config.py:97]: Fair enough. Still I see no purpose for it ...

File: python-lib/config.py (hunk 2 above)
Comment (mklaber) [reply python-lib/config.py:126]: We require clean, consistent code. Consumers should be able to expect that two methods with the same signature would expect the same input (a dataiku schema).

Passing in a list of columns and snowflake types would probably be fine, too. It'd mean the method would have a different signature and no ambiguity about what `originalType` actually means.

Comment (czen88) [reply python-lib/config.py:128]: Ok, in order to be consistent we need to be 100% sure that `REPLACE(data_type, '_', '') AS "originalType"` will be matching to Dataiku datatype for all Snowflake types. I currently do not have this confidence, so would prefer just using Snowflake data types as is. With regards to method signature - it is not specific to Dataiku imho.

File: python-lib/config.py [hunk 3]
```diff
@@ -84,23 +90,21 @@ def get_hdfs_location(dataset_config: Mapping[AnyStr, Any], sf_stage_name: AnySt
...
-    table = tab.split('.')[1]
 
-    sql = f'''
-SELECT column_name AS "name", data_type AS "originalType"
-FROM information_schema.columns
-WHERE table_name = '{table}'
-'''
-    if schema:
-        sql += f" AND table_schema = '{schema}'"
+    tbl_sch_re = re.search(r'("?(?P<schema>[^"]+)"?\.)?"?(?P<table>[^"]+)"?', sf_table_name)
```
[the remark this answers is not in the document]
Comment (mklaber) [reply python-lib/config.py:98]: Regular expressions aren't easy to read, sure. But that's the nature of regular expressions.

The original code didn't work when the input was just `"TABLE_NAME"` without also writing additional checks.

File: python-lib/config.py [hunk 4]
```diff
@@ -51,7 +52,12 @@ def get_table_name(dataset_config: Mapping[AnyStr, Any]) -> AnyStr:
         # e.g., mode == query
         raise ValueError(f'Snowflake connection mode {params["mode"]} is not currently supported.')
 
-    return f'"{params["schema"]}"."{params["table"]}"'.replace('${projectKey}', project_key)
+    table_name = f'"{params["table"]}"'
+
+    if 'schema' in params:
```
[the remark this answers is not in the document]
Comment (mklaber) [reply python-lib/config.py:57]: Always forget about that. Good point.

File: python-lib/config.py [hunk 5]
```diff
@@ -51,7 +52,12 @@ def get_table_name(dataset_config: Mapping[AnyStr, Any]) -> AnyStr:
         # e.g., mode == query
         raise ValueError(f'Snowflake connection mode {params["mode"]} is not currently supported.')
 
-    return f'"{params["schema"]}"."{params["table"]}"'.replace('${projectKey}', project_key)
```
Comment (czen88) [inline python-lib/config.py:54]: Below code looks much nicer imho. You do not need using `if` statement
`return f'"{params.get('schema', '')}"."{params["table"]}"'.replace('${projectKey}', project_key)`

File: python-lib/config.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (czen88) [reply python-lib/config.py:98]: If you want to use Regexp this is fine, but in this case the code needs to document what this expression is doing.
You can put what table formats this will handle for example.

File: python-lib/config.py (hunk 1 above)
Comment (czen88) [reply python-lib/config.py:104]: If you aim to produce Dataiku datatypes from Snowflake datatypes - the code is incomplete. You need to implement mapping for other datatypes as well. Otherwise results of this method are not reliable.
`REPLACE(data_type, '_', '')` does not handle all datatypes imho.

File: python-lib/config.py [hunk 6]
```diff
@@ -51,7 +52,12 @@ def get_table_name(dataset_config: Mapping[AnyStr, Any]) -> AnyStr:
...
+    table_name = f'"{params["table"]}"'
+
+    if params.get('schema', ''):
+        table_name = f'"{params["schema"]}".{table_name}'
```
[the remark this answers is not in the document]
Comment (mklaber) [reply python-lib/config.py:58]: It is. See Line 55. And see associated tests.

File: python-lib/config.py (hunk 5 above)
Comment (mklaber) [reply python-lib/config.py:54]: Again, you do need the if statement. See tests.

**Comment by mklaber** [comment]
I'm not sure why, but GitHub is preventing me from replying to a couple of your comments. Regarding [https://github.com/outrauk/dataiku-plugin-snowflake-hdfs/pull/17#discussion_r462894580](this one):

> If you aim to produce Dataiku datatypes from Snowflake datatypes - the code is incomplete. You need to implement mapping for other datatypes as well. Otherwise results of this method are not reliable.
REPLACE(data_type, '_', '') does not handle all datatypes imho.

Can you provide an example? I'm not sure what you mean—there is no filter on `data_type`, so it will inherently map all datatypes.

File: python-lib/config.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mklaber) [reply python-lib/config.py:100]: I've added comments to describe the regexp. Examples of what it will handle are self documenting in the associated tests (which I've also added additional commentary to).

File: python-lib/config.py (hunk 1 above)
Comment (mklaber) [reply python-lib/config.py:106]: I'm not sure what you mean by this—there are no filters on `data_type` so all are mapped. Do you have an example?

**Comment by czen88** [comment]
> @czen88 I'm not sure why, but GitHub is preventing me from replying to a couple of your comments. Regarding [this one](https://github.com/outrauk/dataiku-plugin-snowflake-hdfs/pull/17#discussion_r462894580):
> 
> > If you aim to produce Dataiku datatypes from Snowflake datatypes - the code is incomplete. You need to implement mapping for other datatypes as well. Otherwise results of this method are not reliable.
> > REPLACE(data_type, '_', '') does not handle all datatypes imho.
> 
> Can you provide an example? I'm not sure what you mean—there is no filter on `data_type`, so it will inherently map all datatypes.

If you run this SQL:
```
SELECT DISTINCT data_type
FROM information_schema.COLUMNS
WHERE table_name = 'PROPERTY_COMBINED'
AND table_schema = 'PUBLIC'
```

you will get these datatypes:
image

This is different what you see in Dataiku for same table:
https://dataiku.outra.co.uk/projects/PROPERTY_CLEAN/datasets/property_combined_sf/settings/#schema

File: python-lib/config.py (hunk 2 above)
Comment (mklaber) [reply python-lib/config.py:130]: RE https://github.com/outrauk/dataiku-plugin-snowflake-hdfs/pull/17#issuecomment-666291831

Okay, understood.

Two options, which do you prefer?

1. Map the additional types (`TEXT` &rarr; `VARCHAR`, `FLOAT` &rarr; `DOUBLE`, and assume there aren't other differences)
2. Pass this as `{"name": "someColumnName", "sfDataType": "TIMESTAMP_NTZ"}` and add the conditional for `sfDataType` in `get_snowflake_to_hdfs_query`

Comment (mklaber) [reply python-lib/config.py]: I went with #2 which allowed me to `git cherry-pick` your PR's commit.

File: python-lib/config.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (czen88) [reply python-lib/config.py:58]: Oh, ok - fair enough. I did not notice it is assigned earlier in code.
